File: polyglot/go/profile_parser.go
package main

import (
	"encoding/json"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// Profile represents a GitHub profile configuration.
type Profile struct {
	Name        string   `json:"name,omitempty"`
	Username    string   `json:"username,omitempty"`
	Bio         string   `json:"bio,omitempty"`
	Location    string   `json:"location,omitempty"`
	Website     string   `json:"website,omitempty"`
	Email       string   `json:"email,omitempty"`
	AvatarURL   string   `json:"avatar_url,omitempty"`
	SocialLinks []SocialLink `json:"social_links,omitempty"`
	Repositories []Repository `json:"repositories,omitempty"`
}

// SocialLink represents a social media link.
type SocialLink struct {
	Name  string `json:"name,omitempty"`
	URL   string `json:"url,omitempty"`
	Type  string `json:"type,omitempty"` // e.g., "twitter", "linkedin"
	Icon  string `json:"icon,omitempty"`
}

// Repository represents a GitHub repository.
type Repository struct {
	Name        string `json:"name,omitempty"`
	Description string `json:"description,omitempty"`
	Stars       int    `json:"stars,omitempty"`
	Forks       int    `json:"forks,omitempty"`
	URL         string `json:"url,omitempty"`
}

// Config represents the full configuration file structure.
type Config struct {
	Version   string     `json:"version,omitempty" yaml:"version,omitempty"`
	Profile   Profile    `json:"profile,omitempty" yaml:"profile,omitempty"`
	Metadata  Metadata   `json:"metadata,omitempty" yaml:"metadata,omitempty"`
}

// Metadata contains optional metadata about the config file.
type Metadata struct {
	Author     string `json:"author,omitempty" yaml:"author,omitempty"`
	Created    string `json:"created,omitempty" yaml:"created,omitempty"`
	LastModified string `json:"last_modified,omitempty" yaml:"last_modified,omitempty"`
}

// DefaultProfile returns a Profile with default values.
func DefaultProfile() *Profile {
	return &Profile{
		Bio:         "Go developer and open source contributor",
		Location:    "San Francisco, CA",
		SocialLinks: []SocialLink{{Name: "GitHub", Type: "github"}},
	}
}

// ValidateEmail checks if the email is in a valid format.
func ValidateEmail(email string) bool {
	if email == "" {
		return true // Empty is allowed
	}
	pattern := `^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`
	matched, _ := regexp.MatchString(pattern, email)
	return matched
}

// NormalizeURL removes trailing slashes and ensures http/https prefix.
func NormalizeURL(rawURL string) (string, error) {
	if rawURL == "" {
		return "", nil
	}
	
	trimmed := strings.TrimSpace(rawURL)
	
	// Add protocol if missing
	if !strings.HasPrefix(trimmed, "http://") && !strings.HasPrefix(trimmed, "https://") {
		trimmed = "https://" + trimmed
	}
	
	// Remove trailing slash (except for root domain)
	if len(trimmed) > 7 && trimmed[len(trimmed)-1] == '/' {
		trimmed = trimmed[:len(trimmed)-1]
	}
	
	return trimmed, nil
}

// ValidateSocialLink checks if a social link is valid.
func ValidateSocialLink(link SocialLink) error {
	if link.URL != "" {
		parsed, err := url.Parse(link.URL)
		if err != nil {
			return fmt.Errorf("invalid URL: %w", err)
		}
		
		if parsed.Scheme == "" || (parsed.Host == "" && parsed.Path == "") {
			return fmt.Errorf("URL must have a host or path")
		}
		
		// Check for common social domains
		validDomains := map[string]bool{
			"github.com": true, "twitter.com": true, "linkedin.com": true,
			"instagram.com": true, "youtube.com": true, "reddit.com": true,
			"discord.com": true, "twitch.tv": true,
		}
		
		if parsed.Host != "" {
			if !validDomains[parsed.Host] && !strings.Contains(parsed.Host, ".") {
				return fmt.Errorf("unknown social domain: %s", parsed.Host)
			}
		}
	}
	
	return nil
}

// ValidateRepository checks if a repository entry is valid.
func ValidateRepository(repo Repository) error {
	if repo.Name == "" {
		return fmt.Errorf("repository name cannot be empty")
	}
	
	if !strings.HasPrefix(repo.URL, "https://github.com/") && 
	   !strings.HasPrefix(repo.URL, "http://github.com/") {
		return fmt.Errorf("URL should point to a GitHub repository")
	}
	
	return nil
}

// ValidateConfig performs comprehensive validation of the config.
func (c *Config) Validate() error {
	var issues []string
	
	// Check profile name
	if c.Profile.Name == "" && c.Profile.Username == "" {
		issues = append(issues, "profile: at least one of 'name' or 'username' is required")
	}
	
	// Check email format if provided
	if c.Profile.Email != "" && !ValidateEmail(c.Profile.Email) {
		issues = append(issues, fmt.Sprintf("profile: invalid email format: %s", c.Profile.Email))
	}
	
	// Validate website URL
	if c.Profile.Website != "" {
		parsed, err := url.Parse(c.Profile.Website)
		if err == nil && parsed.Host != "" {
			c.Profile.Website = parsed.String() // Normalize it
		} else if err != nil {
			issues = append(issues, fmt.Sprintf("profile: invalid website URL: %s", c.Profile.Website))
		}
	}
	
	// Validate social links
	for i, link := range c.Profile.SocialLinks {
		if link.URL == "" && link.Name != "" {
			continue // Name-only is fine
		}
		
		err := ValidateSocialLink(link)
		if err != nil {
			issues = append(issues, fmt.Sprintf("social_links[%d]: %v", i, err))
		}
	}
	
	// Validate repositories
	for i, repo := range c.Profile.Repositories {
		err := ValidateRepository(repo)
		if err != nil {
			issues = append(issues, fmt.Sprintf("repositories[%d]: %v", i, err))
		}
	}
	
	if len(issues) > 0 {
		return fmt.Errorf("validation failed:\n\t%s", strings.Join(issues, "\n\t"))
	}
	
	return nil
}

// ApplyDefaults fills in missing optional fields with sensible defaults.
func (c *Config) ApplyDefaults() {
	if c.Profile.Name == "" && c.Profile.Username == "" {
		c.Profile.Name = "Anonymous"
	}
	
	if c.Profile.Bio == "" {
		c.Profile.Bio = DefaultProfile().Bio
	}
	
	if c.Profile.Location == "" {
		c.Profile.Location = DefaultProfile().Location
	}
	
	// Normalize all URLs
	if err := normalizeURLs(c); err != nil {
		fmt.Printf("Warning: %v\n", err)
	}
}

func normalizeURLs(c *Config) error {
	var normalized []string
	
	if c.Profile.Website != "" {
		n, _ := NormalizeURL(c.Profile.Website)
		c.Profile.Website = n
		normalized = append(normalized, "website")
	}
	
	for i, link := range c.Profile.SocialLinks {
		if link.URL != "" {
			n, err := NormalizeURL(link.URL)
			if err == nil {
				c.Profile.SocialLinks[i].URL = n
				normalized = append(normalized, fmt.Sprintf("social_links[%d]", i))
			} else {
				return err
			}
		}
	}
	
	for i, repo := range c.Profile.Repositories {
		if repo.URL != "" {
			n, _ := NormalizeURL(repo.URL)
			c.Profile.Repositories[i].URL = n
		}
	}
	
	if len(normalized) > 0 {
		fmt.Printf("Normalized %d URL(s)\n", len(normalized))
	}
	
	return nil
}

// ParseConfig reads and parses a configuration file.
func ParseConfig(path string, format string) (*Config, error) {
	var data []byte
	
	// Determine if we need to read the file
	if path == "-" || path == "stdin" {
		data = make([]byte, 0)
		if _, err := fmt.Fscan(&data); err != nil {
			return nil, err
		}
		path = "<stdin>"
	} else if filepath.Ext(path) == ".json" {
		var err error
		data, err = os.ReadFile(path)
		if err != nil {
			return nil, fmt.Errorf("reading JSON file: %w", err)
		}
	} else {
		var err error
		data, err = os.ReadFile(path)
		if err != nil {
			return nil, fmt.Errorf("reading YAML/JSON file: %w", err)
		}
	}
	
	// Try JSON first, then fall back to YAML if needed
	var config Config
	
	// Attempt JSON parsing
	err := json.Unmarshal(data, &config)
	if err == nil {
		return &config, nil
	}
	
	// Fall back to YAML (using a simple parser for common cases)
	config = parseYAMLFallback(string(data))
	return &config, nil
}

// parseYAMLFallback is a minimal YAML parser for common GitHub profile configs.
func parseYAMLFallback(content string) Config {
	var config Config
	
	// Simple key-value extraction for basic YAML support
	lines := strings.Split(strings.TrimSpace(content), "\n")
	
	for _, line := range lines {
		line = strings.TrimSpace(line)
		
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		
		parts := strings.SplitN(line, ":", 2)
		if len(parts) != 2 {
			continue
		}
		
		key := strings.TrimSpace(parts[0])
		value := strings.TrimSpace(parts[1])
		
		// Extract nested keys for profile section
		if key == "profile" {
			config.Profile.Name = extractYAMLValue(value, "name")
			config.Profile.Username = extractYAMLValue(value, "username")
			config.Profile.Bio = extractYAMLValue(value, "bio")
			config.Profile.Location = extractYAMLValue(value, "location")
			config.Profile.Website = extractYAMLValue(value, "website")
			config.Profile.Email = extractYAMLValue(value, "email")
		} else if key == "social_links" {
			// Parse social links array (simplified)
			if strings.HasPrefix(value, "[") && strings.HasSuffix(value, "]") {
				links := parseSocialLinksArray(value)
				config.Profile.SocialLinks = append(config.Profile.SocialLinks, links...)
			} else if value != "" {
				link := SocialLink{URL: value}
				config.Profile.SocialLinks = append(config.Profile.SocialLinks, link)
			}
		} else if key == "repositories" || key == "repos" {
			if strings.HasPrefix(value, "[") && strings.HasSuffix(value, "]") {
				repos := parseRepositoriesArray(value)
				config.Profile.Repositories = append(config.Profile.Repositories, repos...)
			} else if value != "" {
				repo := Repository{Name: value}
				config.Profile.Repositories = append(config.Profile.Repositories, repo)
			}
		}
	}
	
	return config
}

func extractYAMLValue(nested string, field string) string {
	pattern := fmt.Sprintf(`"%s"\s*:\s*"([^"]*)"`+`|`+fmt.Sprintf(`'%s'\s*:\s*'([^']*)'`, field, field)
	matched, _ := regexp.MatchString(pattern, nested)
	if matched {
		return "" // Found it in complex format
	}
	
	// Simple extraction: find "field": "value" pattern
	pattern = fmt.Sprintf(`"%s"\s*:\s*"([^"]*)"`+`|`+fmt.Sprintf(`'%s'\s*:\s*'([^']*)'`, field, field)
	matches := regexp.FindAllStringSubmatch(pattern, -1)
	
	for _, match := range matches {
		if len(match) >= 3 && match[2] != "" {
			return match[2]
		}
	}
	
	return ""
}

func parseSocialLinksArray(arrayStr string) []SocialLink {
	var links []SocialLink
	
	// Simple parser: look for "name": "url" patterns within the array
	pattern := `"\s*name\s*:\s*"([^"]*)"`+`|`+`'\s*name\s*:\s*'([^']*)'`
	matches := regexp.FindAllStringSubmatch(arrayStr, -1)
	
	for _, match := range matches {
		if len(match) >= 3 && (match[2] != "" || match[3] != "") {
			name := strings.TrimSpace(match[2])
			if name == "" {
				name = strings.TrimSpace(match[3])
			}
			
			link := SocialLink{Name: name, Type: "github"}
			links = append(links, link)
		}
	}
	
	return links
}

func parseRepositoriesArray(arrayStr string) []Repository {
	var repos []Repository
	
	pattern := `"\s*name\s*:\s*"([^"]*)"`+`|`+`'\s*name\s*:\s*'([^']*)'`
	matches := regexp.FindAllStringSubmatch(arrayStr, -1)
	
	for _, match := range matches {
		if len(match) >= 3 && (match[2] != "" || match[3] != "") {
			name := strings.TrimSpace(match[2])
			if name == "" {
				name = strings.TrimSpace(match[3])
			}
			
			repo := Repository{Name: name, Description: "Repository"}
			repos = append(repos, repo)
		}
	}
	
	return repos
}

// PrintProfile outputs the profile in a formatted way.
func (p *Profile) Print() {
	fmt.Printf("=== GitHub Profile ===\n")
	fmt.Printf("Name:    %s\n", p.Name)
	if p.Username != "" {
		fmt.Printf("Username:%s\n", p.Username)
	}
	fmt.Printf("Bio:     %s\n", p.Bio)
	fmt.Printf("Location: %s\n", p.Location)
	
	if p.Website != "" {
		fmt.Printf("Website: %s\n", p.Website)
	}
	
	if p.Email != "" {
		fmt.Printf("Email:   %s\n", p.Email)
	}
	
	if p.AvatarURL != "" {
		fmt.Printf("Avatar:  %s\n", p.AvatarURL)
	}
	
	if len(p.SocialLinks) > 0 {
		fmt.Println("\nSocial Links:")
		for _, link := range p.SocialLinks {
			if link.URL != "" {
				fmt.Printf("  • %s: %s\n", link.Name, link.URL)
			} else if link.Name != "" {
				fmt.Printf("  • %s (type: %s)\n", link.Name, link.Type)
			}
		}
	}
	
	if len(p.Repositories) > 0 {
		fmt.Println("\nRepositories:")
		for _, repo := range p.Repositories {
			desc := ""
			if repo.Description != "" {
				desc = fmt.Sprintf(" | %s", repo.Description)
			}
			stars := ""
			forks := ""
			if repo.Stars > 0 {
				stars = fmt.Sprintf(" (⭐ %d)", repo.Stars)
			}
			if repo.Forks > 0 {
				forks = fmt.Sprintf(" (🔀 %d)", repo.Forks)
			}
			fmt.Printf("  • %s

File: polyglot/typescript/profile_parser.ts
import * as fs from 'fs';
import * as yaml from 'yaml';

export interface ProfileConfig {
  user: UserSettings;
  repositories?: Repository[];
  links?: Link[];
  badges?: Badge[];
  metadata?: Metadata;
}

interface UserSettings {
  name: string;
  bio: string;
  location: string;
  avatar_url: string;
  hide_email?: boolean;
  email?: string;
  twitter_username?: string;
  linkedin_username?: string;
  website?: string;
}

interface Repository {
  name: string;
  description: string;
  url: string;
  featured?: boolean;
  pinned?: boolean;
  tags?: string[];
}

interface Link {
  title: string;
  url: string;
  icon?: string;
}

interface Badge {
  name: string;
  description: string;
  url: string;
  color?: string;
}

interface Metadata {
  version: string;
  last_updated: string;
  author: string;
}

export class ProfileParserError extends Error {
  constructor(message: string, public readonly code: string) {
    super(message);
    this.name = 'ProfileParserError';
  }
}

class ProfileParser {
  private static readonly VALID_FIELDS = new Set([
    'name', 'bio', 'location', 'avatar_url', 'hide_email', 
    'email', 'twitter_username', 'linkedin_username', 'website'
  ]);

  constructor(private readonly configPath: string) {}

  parse(): ProfileConfig {
    let rawContent: string;
    
    try {
      rawContent = fs.readFileSync(this.configPath, 'utf-8');
    } catch (error) {
      if (error instanceof Error && error.code === 'ENOENT') {
        throw new ProfileParserError(
          `Config file not found at "${this.configPath}"`,
          'FILE_NOT_FOUND'
        );
      }
      throw new ProfileParserError(`Failed to read config file: ${error}`, 'READ_ERROR');
    }

    if (!rawContent.trim()) {
      throw new ProfileParserError('Config file is empty', 'EMPTY_FILE');
    }

    let parsed: any;
    try {
      parsed = yaml.parse(rawContent);
    } catch (error) {
      throw new ProfileParserError(
        `YAML parsing failed: ${(error as Error).message}`,
        'PARSE_ERROR'
      );
    }

    return this.validateAndNormalize(parsed);
  }

  private validateAndNormalize(data: any): ProfileConfig {
    if (!data.user) {
      throw new ProfileParserError('Missing required "user" section', 'MISSING_USER');
    }

    const user = data.user;
    
    // Validate and normalize user fields
    this.normalizeUser(user);

    // Normalize repositories
    if (data.repositories) {
      data.repositories = this.normalizeRepositories(data.repositories as any[]);
    }

    // Normalize links
    if (data.links) {
      data.links = this.normalizeLinks(data.links as any[]);
    }

    // Normalize badges
    if (data.badges) {
      data.badges = this.normalizeBadges(data.badges as any[]);
    }

    return data;
  }

  private normalizeUser(user: UserSettings): void {
    // Ensure required fields exist
    if (!user.name || user.name.trim() === '') {
      throw new ProfileParserError('User "name" is required and cannot be empty', 'MISSING_NAME');
    }

    if (!user.bio) {
      throw new ProfileParserError('User "bio" is required', 'MISSING_BIO');
    }

    // Validate URL fields
    this.validateUrl(user.avatar_url, 'avatar_url');
    this.validateUrl(user.website, 'website');
    
    // Normalize email format if present
    if (user.email) {
      const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
      if (!emailRegex.test(user.email)) {
        throw new ProfileParserError(`Invalid email format: "${user.email}"`, 'INVALID_EMAIL');
      }
    }

    // Validate social handles
    this.validateHandle(user.twitter_username, 'twitter_username');
    this.validateHandle(user.linkedin_username, 'linkedin_username');
  }

  private validateUrl(field: string | undefined, fieldName: string): void {
    if (field && !/^https?:\/\//i.test(field)) {
      throw new ProfileParserError(
        `Invalid URL format for "${fieldName}": must start with http:// or https://`,
        'INVALID_URL'
      );
    }
  }

  private validateHandle(handle: string | undefined, fieldName: string): void {
    if (handle && !/^[a-zA-Z0-9_]{1,50}$/.test(handle)) {
      throw new ProfileParserError(
        `Invalid handle format for "${fieldName}": must be alphanumeric with underscores`,
        'INVALID_HANDLE'
      );
    }
  }

  private normalizeRepositories(repos: any[]): Repository[] {
    return repos.map((repo, index) => {
      if (!repo.name || !repo.description) {
        throw new ProfileParserError(
          `Repository at index ${index} is missing "name" or "description"`,
          'MISSING_REPO_FIELD'
        );
      }

      this.validateUrl(repo.url, `repository[${index}].url`);

      return {
        name: repo.name.trim(),
        description: repo.description.trim(),
        url: repo.url,
        featured: !!repo.featured,
        pinned: !!repo.pinned,
        tags: Array.isArray(repo.tags) ? repo.tags : []
      };
    });
  }

  private normalizeLinks(links: any[]): Link[] {
    return links.map((link, index) => {
      if (!link.title || !link.url) {
        throw new ProfileParserError(
          `Link at index ${index} is missing "title" or "url"`,
          'MISSING_LINK_FIELD'
        );
      }

      this.validateUrl(link.url, `link[${index}].url`);

      return {
        title: link.title.trim(),
        url: link.url,
        icon: link.icon || undefined
      };
    });
  }

  private normalizeBadges(badges: any[]): Badge[] {
    return badges.map((badge, index) => {
      if (!badge.name || !badge.description || !badge.url) {
        throw new ProfileParserError(
          `Badge at index ${index} is missing required fields`,
          'MISSING_BADGE_FIELD'
        );
      }

      this.validateUrl(badge.url, `badge[${index}].url`);

      return {
        name: badge.name.trim(),
        description: badge.description.trim(),
        url: badge.url,
        color: badge.color || undefined
      };
    });
  }

  static fromFile(path: string): ProfileConfig {
    const parser = new ProfileParser(path);
    return parser.parse();
  }
}

// Export a default instance for convenience
export const profileParser = new ProfileParser('.github/profile.yml');

// Demo / Entry Point
if (require.main === module) {
  async function main() {
    try {
      console.log('GitHub Profile Parser - Demo\n');
      console.log(`Looking for config at: ${profileParser.configPath}\n`);

      const config = profileParser.parse();

      console.log('Parsed Configuration:\n');
      console.log(JSON.stringify(config, null, 2));

      // Quick summary
      console.log('\n--- Summary ---');
      console.log(`Name: ${config.user.name}`);
      console.log(`Bio: ${config.user.bio}`;
      console.log(`Location: ${config.user.location}`);
      console.log(`Repos configured: ${config.repositories?.length || 0}`);
      console.log(`Links configured: ${config.links?.length || 0}`);
      console.log(`Badges configured: ${config.badges?.length || 0}`);

    } catch (error) {
      if (error instanceof ProfileParserError) {
        console.error(`Error (${error.code}): ${error.message}`);
        process.exit(1);
      } else {
        console.error('Unexpected error:', error);
        process.exit(1);
      }
    }
  }

  main();
}

export default ProfileParser;